File: examples/elixir/test_client.exs
#!/usr/bin/env elixir

# Example Elixir client for testing Mock PVE API Server.
#
# This script demonstrates how to use the mock server for testing
# Elixir-based PVE client libraries.
#
# Requirements:
#   mix deps.get
#
# Usage:
#   # Start mock server first
#   podman run -d -p 8006:8006 docker.io/jrjsmrtn/mock-pve-api:latest
#   
#   # Run this script
#   elixir test_client.exs

defmodule MockPVEClient do
  @moduledoc """
  Simple PVE client for testing against mock server.
  """

  @base_url "http://localhost:8006/api2/json"

  def get(endpoint, params \\ []) do
    url = build_url(endpoint, params)
    
    case HTTPoison.get(url, [], timeout: 10_000) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
        {:ok, Jason.decode!(body)}
      
      {:ok, %HTTPoison.Response{status_code: status, body: body}} ->
        {:error, {status, body}}
      
      {:error, reason} ->
        {:error, reason}
    end
  end

  def post(endpoint, data \\ %{}) do
    url = build_url(endpoint)
    body = Jason.encode!(data)
    headers = [{"Content-Type", "application/json"}]
    
    case HTTPoison.post(url, body, headers, timeout: 10_000) do
      {:ok, %HTTPoison.Response{status_code: 200, body: response_body}} ->
        {:ok, Jason.decode!(response_body)}
      
      {:ok, %HTTPoison.Response{status_code: status, body: response_body}} ->
        {:error, {status, response_body}}
      
      {:error, reason} ->
        {:error, reason}
    end
  end

  defp build_url(endpoint, params \\ []) do
    endpoint = String.trim_leading(endpoint, "/")
    base = "#{@base_url}/#{endpoint}"
    
    case params do
      [] -> base
      params -> "#{base}?#{URI.encode_query(params)}"
    end
  end
end

defmodule TestRunner do
  @moduledoc """
  Test runner for Mock PVE API Server.
  """

  def run do
    IO.puts("🚀 Mock PVE API Client Test (Elixir)")
    IO.puts(String.duplicate("=", 40))
    
    try do
      version_data = test_version_info()
      test_cluster_status()
      test_nodes_list()
      test_cluster_resources()
      test_storage_content()
      test_resource_pools()
      test_version_specific_features(version_data)
      
      IO.puts("\n" <> String.duplicate("=", 40))
      IO.puts("🎉 All tests completed successfully!")
      IO.puts("\nMock PVE API Server is working correctly.")
      
    rescue
      error ->
        case error do
          %HTTPoison.Error{reason: :econnrefused} ->
            IO.puts("❌ Error: Could not connect to Mock PVE API Server")
            IO.puts("\nMake sure the server is running:")
            IO.puts("  podman run -d -p 8006:8006 docker.io/jrjsmrtn/mock-pve-api:latest")
            System.halt(1)
          
          _ ->
            IO.puts("❌ Unexpected error: #{inspect(error)}")
            System.halt(1)
        end
    end
  end

  defp test_version_info do
    IO.puts("🔍 Testing version information...")
    
    {:ok, response} = MockPVEClient.get("version")
    version_data = response["data"]
    
    IO.puts("  ✅ PVE Version: #{version_data["version"] || "Unknown"}")
    IO.puts("  ✅ Release: #{version_data["release"] || "Unknown"}")
    IO.puts("  ✅ Repository ID: #{version_data["repoid"] || "Unknown"}")
    
    version_data
  end

  defp test_cluster_status do
    IO.puts("\n🔍 Testing cluster status...")
    
    {:ok, response} = MockPVEClient.get("cluster/status")
    
    Enum.each(response["data"], fn item ->
      case item["type"] do
        "node" ->
          IO.puts("  ✅ Node: #{item["name"]} - Status: #{item["status"]}")
        
        "cluster" ->
          IO.puts("  ✅ Cluster: #{item["name"]}")
        
        _ ->
          :ok
      end
    end)
    
    response["data"]
  end

  defp test_nodes_list do
    IO.puts("\n🔍 Testing nodes list...")
    
    {:ok, response} = MockPVEClient.get("nodes")
    
    Enum.each(response["data"], fn node ->
      cpu_percent = (node["cpu"] || 0) * 100 |> Float.round(1)
      memory_gb = (node["mem"] || 0) / :math.pow(1024, 3) |> Float.round(1)
      
      IO.puts("  ✅ Node: #{node["node"]} - Status: #{node["status"]}")
      IO.puts("     CPU: #{cpu_percent}% - Memory: #{memory_gb}GB")
    end)
    
    response["data"]
  end

  defp test_cluster_resources do
    IO.puts("\n🔍 Testing cluster resources...")
    
    {:ok, response} = MockPVEClient.get("cluster/resources")
    
    resource_counts = 
      response["data"]
      |> Enum.reduce(%{}, fn resource, acc ->
        type = resource["type"]
        Map.put(acc, type, Map.get(acc, type, 0) + 1)
      end)
    
    Enum.each(response["data"], fn resource ->
      case resource["type"] do
        "qemu" ->
          IO.puts("  ✅ VM #{resource["vmid"]}: #{resource["name"]} - Status: #{resource["status"]}")
        
        "lxc" ->
          IO.puts("  ✅ CT #{resource["vmid"]}: #{resource["name"]} - Status: #{resource["status"]}")
        
        "storage" ->
          plugintype = resource["plugintype"] || "unknown"
          IO.puts("  ✅ Storage: #{resource["storage"]} - Type: #{plugintype}")
        
        _ ->
          :ok
      end
    end)
    
    IO.puts("\n  📊 Resource Summary:")
    Enum.each(resource_counts, fn {type, count} ->
      IO.puts("     #{type}: #{count}")
    end)
    
    response["data"]
  end

  defp test_storage_content(node \\ "pve-node-1", storage \\ "local") do
    IO.puts("\n🔍 Testing storage content for #{storage} on #{node}...")
    
    case MockPVEClient.get("nodes/#{node}/storage/#{storage}/content") do
      {:ok, response} ->
        content_types = 
          response["data"]
          |> Enum.reduce(%{}, fn item, acc ->
            type = item["content"]
            Map.put(acc, type, Map.get(acc, type, 0) + 1)
          end)
        
        Enum.each(response["data"], fn item ->
          size_gb = (item["size"] || 0) / :math.pow(1024, 3) |> Float.round(2)
          IO.puts("  ✅ #{item["volid"]} - Type: #{item["content"]} - Size: #{size_gb}GB")
        end)
        
        IO.puts("\n  📊 Content Summary:")
        Enum.each(content_types, fn {type, count} ->
          IO.puts("     #{type}: #{count}")
        end)
      
      {:error, reason} ->
        IO.puts("  ⚠️  Storage content test failed: #{inspect(reason)}")
    end
  end

  defp test_resource_pools do
    IO.puts("\n🔍 Testing resource pools...")
    
    case MockPVEClient.get("pools") do
      {:ok, response} ->
        Enum.each(response["data"], fn pool ->
          comment = pool["comment"] || "No comment"
          IO.puts("  ✅ Pool: #{pool["poolid"]} - Comment: #{comment}")
          
          if pool["members"] do
            IO.puts("     Members: #{length(pool["members"])}")
          end
        end)
        
        response["data"]
      
      {:error, reason} ->
        IO.puts("  ⚠️  Resource pools test failed: #{inspect(reason)}")
        []
    end
  end

  defp test_version_specific_features(version_data) do
    version = version_data["version"] || "8.3"
    IO.puts("\n🔍 Testing version-specific features for PVE #{version}...")
    
    # Test SDN endpoints (available in PVE 8.0+)
    if Version.compare(version, "8.0") != :lt do
      test_sdn_features()
    else
      IO.puts("  ⏭️  SDN features not available in PVE #{version}")
    end
    
    # Test backup providers (available in PVE 8.2+)
    if Version.compare(version, "8.2") != :lt do
      test_backup_providers()
    else
      IO.puts("  ⏭️  Backup providers not available in PVE #{version}")
    end
  end

  defp test_sdn_features do
    try do
      IO.puts("  🔍 Testing SDN zones (PVE 8.0+)...")
      {:ok, zones_response} = MockPVEClient.get("cluster/sdn/zones")
      IO.puts("  ✅ SDN Zones: #{length(zones_response["data"])} zones available")
      
      IO.puts("  🔍 Testing SDN vnets (PVE 8.0+)...")
      {:ok, vnets_response} = MockPVEClient.get("cluster/sdn/vnets")
      IO.puts("  ✅ SDN VNets: #{length(vnets_response["data"])} vnets available")
      
    rescue
      error ->
        IO.puts("  ⚠️  SDN endpoints test failed: #{inspect(error)}")
    end
  end

  defp test_backup_providers do
    try do
      IO.puts("  🔍 Testing backup providers (PVE 8.2+)...")
      {:ok, providers_response} = MockPVEClient.get("cluster/backup-info/providers")
      IO.puts("  ✅ Backup Providers: #{length(providers_response["data"])} providers available")
      
    rescue
      error ->
        IO.puts("  ⚠️  Backup providers test failed: #{inspect(error)}")
    end
  end
end

# Check if required dependencies are available
dependencies = [:httpoison, :jason]

missing_deps = Enum.filter(dependencies, fn dep ->
  try do
    dep.__info__(:module)
    false
  rescue
    _ -> true
  end
end)

if length(missing_deps) > 0 do
  IO.puts("❌ Missing dependencies: #{Enum.join(missing_deps, ", ")}")
  IO.puts("Please install them with: mix deps.get")
  System.halt(1)
else
  # All dependencies available, run tests
  TestRunner.run()
end

File: examples/python/test_client.py
#!/usr/bin/env python3
"""
Example Python client for testing Mock PVE API Server.

This script demonstrates how to use the mock server for testing
Python-based PVE client libraries.

Requirements:
    pip install requests

Usage:
    # Start mock server first
    podman run -d -p 8006:8006 docker.io/jrjsmrtn/mock-pve-api:latest
    
    # Run this script
    python test_client.py
"""

import requests
import json
import sys
from urllib3.exceptions import InsecureRequestWarning

# Disable SSL warnings for mock server
requests.urllib3.disable_warnings(InsecureRequestWarning)


class MockPVEClient:
    """Simple PVE client for testing against mock server."""
    
    def __init__(self, host='localhost', port=8006, verify_ssl=False):
        self.base_url = f"http://{host}:{port}/api2/json"
        self.verify_ssl = verify_ssl
        self.session = requests.Session()
    
    def get(self, endpoint, params=None):
        """Make GET request to PVE API."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        response = self.session.get(url, params=params, verify=self.verify_ssl)
        response.raise_for_status()
        return response.json()
    
    def post(self, endpoint, data=None):
        """Make POST request to PVE API."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        response = self.session.post(url, data=data, verify=self.verify_ssl)
        response.raise_for_status()
        return response.json()


def test_version_info(client):
    """Test version information endpoint."""
    print("🔍 Testing version information...")
    response = client.get('version')
    
    version_data = response['data']
    print(f"  ✅ PVE Version: {version_data.get('version')}")
    print(f"  ✅ Release: {version_data.get('release')}")
    print(f"  ✅ Repository ID: {version_data.get('repoid')}")
    
    return version_data


def test_cluster_status(client):
    """Test cluster status endpoint."""
    print("\n🔍 Testing cluster status...")
    response = client.get('cluster/status')
    
    for node in response['data']:
        if node['type'] == 'node':
            print(f"  ✅ Node: {node['name']} - Status: {node['status']}")
        elif node['type'] == 'cluster':
            print(f"  ✅ Cluster: {node['name']}")
    
    return response['data']


def test_nodes_list(client):
    """Test nodes listing endpoint."""
    print("\n🔍 Testing nodes list...")
    response = client.get('nodes')
    
    for node in response['data']:
        print(f"  ✅ Node: {node['node']} - Status: {node['status']}")
        print(f"     CPU: {node.get('cpu', 0)*100:.1f}% - Memory: {node.get('mem', 0)/(1024**3):.1f}GB")
    
    return response['data']


def test_cluster_resources(client):
    """Test cluster resources endpoint."""
    print("\n🔍 Testing cluster resources...")
    response = client.get('cluster/resources')
    
    resource_counts = {}
    for resource in response['data']:
        resource_type = resource['type']
        resource_counts[resource_type] = resource_counts.get(resource_type, 0) + 1
        
        if resource_type == 'qemu':
            print(f"  ✅ VM {resource['vmid']}: {resource['name']} - Status: {resource['status']}")
        elif resource_type == 'lxc':
            print(f"  ✅ CT {resource['vmid']}: {resource['name']} - Status: {resource['status']}")
        elif resource_type == 'storage':
            print(f"  ✅ Storage: {resource['storage']} - Type: {resource.get('plugintype')}")
    
    print(f"\n  📊 Resource Summary:")
    for res_type, count in resource_counts.items():
        print(f"     {res_type}: {count}")
    
    return response['data']


def test_storage_content(client, node='pve-node-1', storage='local'):
    """Test storage content endpoint."""
    print(f"\n🔍 Testing storage content for {storage} on {node}...")
    try:
        response = client.get(f'nodes/{node}/storage/{storage}/content')
        
        content_types = {}
        for item in response['data']:
            content_type = item['content']
            content_types[content_type] = content_types.get(content_type, 0) + 1
            print(f"  ✅ {item['volid']} - Type: {content_type} - Size: {item.get('size', 0)/(1024**3):.2f}GB")
        
        print(f"\n  📊 Content Summary:")
        for content_type, count in content_types.items():
            print(f"     {content_type}: {count}")
            
    except requests.exceptions.HTTPError as e:
        print(f"  ⚠️  Storage content test failed: {e}")


def test_version_specific_features(client, version_data):
    """Test version-specific features."""
    version = version_data.get('version', '8.3')
    print(f"\n🔍 Testing version-specific features for PVE {version}...")
    
    # Test SDN endpoints (available in PVE 8.0+)
    if version >= '8.0':
        try:
            print("  🔍 Testing SDN zones (PVE 8.0+)...")
            response = client.get('cluster/sdn/zones')
            print(f"  ✅ SDN Zones: {len(response['data'])} zones available")
            
            print("  🔍 Testing SDN vnets (PVE 8.0+)...")
            response = client.get('cluster/sdn/vnets')
            print(f"  ✅ SDN VNets: {len(response['data'])} vnets available")
            
        except requests.exceptions.HTTPError as e:
            print(f"  ⚠️  SDN endpoints test failed: {e}")
    else:
        print(f"  ⏭️  SDN features not available in PVE {version}")
    
    # Test backup providers (available in PVE 8.2+)
    if version >= '8.2':
        try:
            print("  🔍 Testing backup providers (PVE 8.2+)...")
            response = client.get('cluster/backup-info/providers')
            print(f"  ✅ Backup Providers: {len(response['data'])} providers available")
        except requests.exceptions.HTTPError as e:
            print(f"  ⚠️  Backup providers test failed: {e}")
    else:
        print(f"  ⏭️  Backup providers not available in PVE {version}")


def main():
    """Main test function."""
    print("🚀 Mock PVE API Client Test")
    print("=" * 40)
    
    # Initialize client
    client = MockPVEClient(host='localhost', port=8006)
    
    try:
        # Test basic endpoints
        version_data = test_version_info(client)
        test_cluster_status(client)
        test_nodes_list(client)
        test_cluster_resources(client)
        test_storage_content(client)
        
        # Test version-specific features
        test_version_specific_features(client, version_data)
        
        print("\n" + "=" * 40)
        print("🎉 All tests completed successfully!")
        print("\nMock PVE API Server is working correctly.")
        
    except requests.exceptions.ConnectionError:
        print("❌ Error: Could not connect to Mock PVE API Server")
        print("\nMake sure the server is running:")
        print("  podman run -d -p 8006:8006 docker.io/jrjsmrtn/mock-pve-api:latest")
        sys.exit(1)
        
    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP Error: {e}")
        sys.exit(1)
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)


if __name__ == '__main__':
    main()